File: Commercial/Java_Commercial_Controller/src/tech/rocketelevators/Battery.java
package tech.rocketelevators;

import java.util.ArrayList;
import java.util.stream.Collectors;

public class Battery {
	Building building;
	int firstFloor;
	ArrayList<Column> columnList;
	int totalColumns;
	int totalElevators;
	boolean isPoweredOn = false;

	public Battery(Building building, int totalColumns, int totalElevators) {
		this.building = building;
		this.totalColumns = totalColumns;
		this.totalElevators = totalElevators;
		this.columnList = new ArrayList<>();
		this.createColumns();

	}

	// makes columns and assigns each column to the floors it'll handle. one column has the basement and first floor, others handle an even share of floors.

	public void createColumns() {
		for (int i = 0; i < totalColumns; i++) {
			int floorsPerColumn = (building.floors - building.basements) / (totalColumns - 1);
			int elevatorsPerColumn = totalElevators / totalColumns;
			switch (i) {
			case 0:
				this.columnList.add(new Column(this, i, elevatorsPerColumn, -building.basements, 0));
				break;

			case 1:
				this.columnList.add(new Column(this, i, elevatorsPerColumn, 0, floorsPerColumn));
				break;

			default:
				int minimum = this.columnList.get(i - 1).maximumFloor + 1;
				int maximum = this.columnList.get(i - 1).maximumFloor + floorsPerColumn;
				this.columnList.add(new Column(this, i, elevatorsPerColumn, minimum, maximum));
				break;
			}
		}
	}

	public Column findColumn(int requestLocation) {
		return this.columnList.stream()
				.filter(c -> requestLocation >= c.minimumFloor && requestLocation <= c.maximumFloor)
				.collect(Collectors.toList()).get(0);

	}

	public void showColumns() {
		for (Column column : columnList) {
			column.columnInfo();
		}
	}

	public void togglePower() {
		isPoweredOn = !isPoweredOn;
		String message = isPoweredOn ? "Power On" : "Power Off";
		System.out.println("Battery: " + message);
	}
}

File: Commercial/TypeScript_Commercial_Controller/Battery.ts
import Building from './Building';
import Column from './Column';

export default class Battery {
    building: Building;
    columnList: Array<Column>;
    totalColumns: number;
    totalElevators: number;
    Scenario1: any;
    Scenario2: any;
    Scenario3: any;
    Scenario4: any;

    constructor(building: Building, columnNumber: number, totalElevators: number) {
        this.building = building;
        this.columnList = [];
        this.totalColumns = columnNumber;
        this.totalElevators = totalElevators;
        this.createColumns();
    }

    createColumns(): void {
        let floorsPerColumn = (this.building.floors - this.building.basements) / (this.totalColumns - 1);
        let elevatorsPerColumn = this.totalElevators / this.totalColumns;
        let minimum = floorsPerColumn;
        let maximum = minimum + floorsPerColumn;
        for (let i = 0; i < this.totalColumns; i++) {
            if (i === 0) {
                this.columnList.push(new Column(i, elevatorsPerColumn, -this.building.basements, 0));
            } else if (i === 1) {
                this.columnList.push(new Column(i, elevatorsPerColumn, 0, floorsPerColumn));
            } else if (i > 1) {
                this.columnList.push(new Column(i, elevatorsPerColumn, minimum + 1, maximum));
                minimum += floorsPerColumn;
                maximum += floorsPerColumn;
            }
        }
    }

    findColumn(requestLocation: number): Column {
        return this.columnList.filter((c) => requestLocation >= c.minFloor && requestLocation <= c.maxFloor)[0];
    }
}
